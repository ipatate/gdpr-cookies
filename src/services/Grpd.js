// @flow
import merge from 'deepmerge';
import GrpdCookie from './GrpdCookie';

export default class Grpd {
  options: OptionsGrpd = {
    name: 'grpd_cookie',
    keepCookies: ['plop'],
    types: ['ads', 'stats'],
  };
  cookie: GrpdCookie;
  activated: Object;

  /**
   * @description constructor !
   * @param {object} options
   * @return {void}
   */
  constructor(options: ?OptionsGrpd = {name: 'grpd_cookie'}): void {
    this.options = merge.all([this.options, options]);
    this.cookie = new GrpdCookie(this.options.name);
    // init activated object and save in cookie
    this.initCookie();
  }

  /**
   * @description get value in options by name
   * @param {string} name
   * @return {void}
   */
  getOption(name: string): void {
    if (this.options[name] !== undefined) {
      return this.options[name];
    }
    return undefined;
  }

  /**
   * @description init cookie settings if not exist
   * @return {void}
   */
  initCookie(): void {
    const _c = this.cookie.getCookie();
    const value = this.createActivatedObject(_c);
    this.cookie.updateCookie(value);
  }

  /**
   * @description create object activated with old value if exist
   * @param {object | void} _c
   * @return {object}
   */
  createActivatedObject(_c: Object | void): Object {
    const {types} = this.options;
    // old value in cookie if exist
    const oldActivated = _c !== undefined ? _c : {};
    const newActivated = {};
    // if type is array
    if (types !== undefined && Array.isArray(types)) {
      types.forEach(name => {
        // old value or true by default
        newActivated[name] =
          oldActivated[name] !== undefined ? oldActivated[name] : true;
      });
    }
    this.activated = newActivated;
    return this.activated;
  }
}
