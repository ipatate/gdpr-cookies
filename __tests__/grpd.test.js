import Cookie from '../src/utils/Cookie.js';
import Grpd from '../src/Grpd.js';

test('test get option value', () => {
  const grpd = new Grpd({name: 'lol'});
  const optionName = grpd.getOption('name');
  expect(optionName).toBe('lol');
});

test('test cookie rgpd exist', () => {
  const grpd = new Grpd();
  const optionName = grpd.getOption('name');
  const _cookie = new Cookie();
  _cookie.remove(optionName);
  expect(grpd.getCookie()).toBeUndefined();
});

test('verify value default grpd cookie', () => {
  const grpd = new Grpd();
  expect(grpd.createActivatedObject()).toMatchSnapshot();
});

test('verify update value with old setting grpd cookie', () => {
  const grpd = new Grpd({type: ['ads', 'stats']});
  const activated = grpd.createActivatedObject({ads: false, foo: 'lol'});
  expect(activated.ads).toBeFalsy();
});

test('update value default grpd cookie', () => {
  const grpd = new Grpd();
  const t = grpd.updateCookie({lol: true, plop: false});
  expect(grpd.getCookie()).toMatchSnapshot();
});

test('get value return undefined', () => {
  const grpd = new Grpd();
  const optionName = grpd.getOption('name');
  const _cookie = new Cookie();
  _cookie.remove(optionName);
  expect(grpd.getCookie()).toBeUndefined();
});
