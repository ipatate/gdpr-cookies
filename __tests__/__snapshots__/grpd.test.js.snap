// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verify value default grpd cookie 1`] = `
Object {
  "ads": true,
  "stats": true,
}
`;
